Add nullbits Bit-C PRO board

jaygreco opened the pull request:
Add the upcoming nullbits Bit-C PRO RP2040 board. Tested and working on our engineering samples.

Relies on this pico-sdk PR: https://github.com/raspberrypi/pico-sdk/pull/1149

micropython-media PR for the board image: https://github.com/micropython/micropython-media/pull/31

I also took the approach that the Weact Studio boards use, and defined the built-in LEDs in a board module so that it's easier for our users to set the RGB LED.

jaygreco committed:
rp2/boards/NULLBITS_BIT_C_PRO: Add Bit-C PRO board.

iabdalkader commented:
When https://github.com/micropython/micropython/pull/9731 is merged, you'll be able to define those LEDs in `pins.csv` just like other ports. @dpgeorge it might be a good time to revisit https://github.com/micropython/micropython/pull/9731 ?

dpgeorge commented:
@jaygreco please rebase on latest master and add a `pins.csv`, now that #9731 is merged.

Also, if the pico-sdk PR is not merged, it's possible to put all that config here in this repo (in the board's directory) and refer to it with cmake variables.

When will this board be available to buy?

jaygreco commented:
Hi @dpgeorge, thanks for the feedback! I rebased against the latest and updated with an additional commit to include the LED pin definitions in `pins.csv`. Please excuse the extra try it took to get the formatting checks to pass 🤦 I also added the board sdk definitions, as I haven't heard any movement on the pico-sdk PR. 

The board will be up for sale by the end of the month. We have been distributing engineering samples and the final boards are set to arrive this week. Happy to share one if it would be useful for you and/or you'd be interested (although it might take a while to get to you!). Let me know if there are additional changes you'd like and if you'd prefer me to squash the commits into one. Thanks!

dpgeorge commented on ports/rp2/boards/NULLBITS_BIT_C_PRO/README.md:
````diff
@@ -0,0 +1,17 @@
+Example:
+
+```python
+> import board
````
Python has `>>>` for the (default) prompt.

dpgeorge commented on ports/rp2/boards/NULLBITS_BIT_C_PRO/modules/board.py:
```diff
@@ -0,0 +1,5 @@
+
+led_r = Pin("LEDR", Pin.OUT, value=1)
+led_g = Pin("LEDG", Pin.OUT, value=1)
+led_b = Pin("LEDB", Pin.OUT, value=1)
```
If you use `machine.Signal(..., invert=True)` then the on/off methods will work in the correct manner.

Also, I suggest to use full colour names here, and match the pin name to the object name, eg:
```py
led_red = Signal("LED_RED", Pin.OUT, invert=True, value=0)
```
(this will require updating pins.csv as well)

This is for consistency, eg to match `ports/stm32/boards/PYBV10/pins.csv`.

dpgeorge replied:
Also, I suggest adding `del Pin` at the end of this file so that `board.Pin` isn't confusingly exposed to the user.

jaygreco replied:
Good feedback! Definitely cleaner. I'll do that.

dpgeorge commented on ports/rp2/boards/NULLBITS_BIT_C_PRO/manifest.py:
```diff
@@ -0,0 +1,2 @@
+include("$(PORT_DIR)/boards/manifest.py")
+freeze("./modules")
```
Please put `board.py` alongside `board.md` (ie not in `modules/` subdir), then this line becomes:
```py
module("board.py")
```

dpgeorge commented on ports/rp2/boards/NULLBITS_BIT_C_PRO/nullbits_bit_c_pro.h:
```diff
@@ -0,0 +1,96 @@
+#endif
+
+#ifndef PICO_DEFAULT_SPI_CSN_PIN
+#define PICO_DEFAULT_SPI_CSN_PIN 21
```
Out of curiosity, why are these defaults they way they are, ie why not the same as on the Pico board?

jaygreco replied:
Routing these IOs to the same location as the Pro Micro SPI pins was easer vs using the same ones on the Pico.

dpgeorge commented on ports/rp2/boards/NULLBITS_BIT_C_PRO/mpconfigboard.h:
```diff
@@ -0,0 +1,9 @@
+// https://nullbits.co/bit-c-pro
+
+#define MICROPY_HW_BOARD_NAME       "nullbits Bit-C PRO"
+#define MICROPY_HW_FLASH_STORAGE_BYTES (PICO_FLASH_SIZE_BYTES - (1 * 512 * 1024)) // 512K reserved
```
Do you want a bigger filesystem?  I'd suggest 3MiB (same as PIMORONI_PICOLIPO_4MB).

jaygreco replied:
With the current configuration, doesn't that leave more room for the filesystem? Thonny reports 3.5MB free.

dpgeorge replied:
Oh, I misread this as leaving only 512k for the filesystem.

So with the current configuration you have 512k for the firmware, and 3.5MiB for the filesystem.

The firmware is currently about 310k, so there is a fair bit of space left to get up to 512k.  So it should be OK to leave this as-is (ie 3.5MiB filesystem).

jaygreco replied:
awesome, thanks for confirming! I wasn't 100% sure.

dpgeorge commented:
This is blocked on `https://nullbits.co/bit-c-pro` existing.

jaygreco commented:
> This is blocked on https://nullbits.co/bit-c-pro existing.

Just went live :)

jaygreco commented:
Thanks for all the feedback and the quick reviews, @dpgeorge! It's been a pleasure.
